### PR TITLE
Show a preview of comments in listings

### PR DESCRIPTION
**Pull request by ciex**
First of all, I am not familiar with the C4 contribution process, please be patient with me :) I just built this for myself today because I wanted to get familiar with the Oasis codebase and wanted to have this feature myself. If it's not in your vision for Oasis, no problem.

**What's the problem you solved?**

On a large monitor, Oasis is only using a fraction of the horizontal space. Also, I was missing Patchworks feature to show a preview of comments on each post in the timeline.

Example:

image

**What solution are you recommending?**

Here I have added a preview of comments on each post in the timeline, which is shown to the side of the original post on large screens and below on small screens (as in Patchwork).

Examples:

image

image

**Caveats**

- The popular page now takes a long time to load (56sec vs 12sec before). Both times are too long to be usable though. I wonder what would be a good way to speed up loading of the page but I feel like I don't know enough about the architecture and also that this would be out of scope for this maybe (?).
- I duplicated a bit of code here that I would like to refactor before this is mergeable. I wanted to hear whether this approach is even wanted before doing the work though. Hope you understand!
- I had a problem committing my changes when Git's commit hook would ask me to run prettier on the code first. As there is no prettier config in the project, that would change all of the rest of the codebase, which is probably not what is intended. So I suggest to add a `.prettierrc.js`.

### CODE REVIEW

File: src/assets/style.css [hunk 1]
```diff
@@ -495,8 +522,7 @@ hr {
     margin-bottom: var(--whole);
   }
   main {
-    width: 100%;
-    max-width: var(--measure);
+    width: 85vw;
```
Comment (ciex) [inline src/assets/style.css]: Need to change this to something that doesn't conflict with the nav bar on small screens

File: src/views/index.js [hunk 2]
```diff
@@ -56,6 +56,8 @@ exports.setLanguage = language => {
 const markdownUrl = "https://commonmark.org/help/";
 const doctypeString = "<!DOCTYPE html>";
 
+const THREAD_PREVIEW_LENGTH = 3;
```
Comment (ciex) [inline src/views/index.js]: I didn't see a configuration file for constants like this so I added it inside the module

File: src/views/index.js [hunk 3]
```diff
@@ -121,6 +123,188 @@ const template = (...elements) => {
 
   return result;
 };
+
+const postInAside = msg => {
```
Comment (ciex) [inline src/views/index.js]: so this is duplicating a lot of code from `post()`. My approach here would be to add a library with utilites for working with post objects to solve things like assembling the object of URL's for a post.

Comment (christianbundy) [reply src/views/index.js]: This would be cool! I've been thinking that maybe the router (`src/index.js`) should be handing these URL builders, and I think koa-router even supports this.

**Comment by christianbundy** [comment]
Hey, thanks for this patch! It's really cool to see you diving into the code and making changes. :rocket: I think it might make sense to spend a bit more time discussing the problems to make sure we agree on what the problem(s) are and how we want to solve them. As I'm sure you've seen, the code is sort of bleh and I know there are some changes that will touch lots of the same code that you touched.

Right now we're in this weird transitional page where we have a page for every type of view, whereas we should probably have filter/view options [like this](https://github.com/fraction/oasis/issues/170#issuecomment-583126856). I think it'd be really really cool to have a thread preview, or "thread rollup" as it's called in Patchwork. I know it isn't ideal, but maybe we could experiment with this by making a new page called "Summaries" or something? I'm imagining it having the same data as "Topics", but we could do a different page if you preferred. Maybe others would have a good idea on how to do this best? I'd love if you could open an issue about summaries so we could all discuss and figure out what we want in a solution. I imagine it'd look a lot like what you've built. :heart: For example, you might have already written some of the code that would resolve #220!

Regarding the thread redesign to target wide screens, what happens when a comment is longer than the original post? I'd love to get feedback from others on this solution, since it isn't something I've seen on other social media, but maybe my initial reaction would change after playing around with it a bit more. I think I'd be really happy with this sort of design if we had an email-style interface where the left side was a summary and the right side was the thread, as discussed [here](https://github.com/fraction/oasis/issues/126#issuecomment-581035222), but I think it's something I'd like to discuss more first to make sure we agree on a problem and have a solution that fits.

> I had a problem committing my changes when Git's commit hook would ask me to run prettier on the code first.

`npm run fix` should solve this, but it should be documented! Or we should use lint-staged or something so this happens automatically...

**Comment by jedahan** [comment]
I really like this experimentation, and take it as a good example that oasis-as-a-platform is in a healthy place to encourage new ideas.

I wonder how we feel about making it easier to incorporate more radical experiments within oasis itself. Like an experiments or labs setting to allow for people to try different things. Just my initial thoughts.

**Comment by jedahan** [comment]
And more specifically on this two-dimensional design, I think its a real cool way of taking advantage of real estate in a novel manner, and differentiating from the monotony of 'getting through a single infinite feed', encouraging looking at the conversation a bit more deeply and with good context.

**Comment by cinnamon-bun** [comment]
Thanks for this experiment!  It's a promising idea - I miss Patchwork's way of showing threads.

❗️ Rambling thoughts ahead ❗️ 

I've been trying to figure out how all these different view styles will encourage or discourage healthy community, and I'm feeling stumped.

### The kind of interaction I'd like to encourage

The feeling I like best is long conversation threads where each message is a paragraph or more and the conversation goes back and forth for a long time.  People engaging with each other deeply, like pen pals.  This happens sometimes on Hacker News though it's more idea-based instead of person-based; on Metafilter; and in Patchwork.  Livejournal also used to feel this way at times.  Github issue conversations are another good example.

Anti-examples are Twitter and Mastodon (too short and scattered), Reddit and Youtube (full of low-effort memes), and Facebook (sometimes long posts, but short replies that don't build on each other).

Oasis has a Twitter-like feel now, where each comment is scattered around instead of seen in context, and I find it confusing.  I prefer Patchwork's thread-centric views.  It makes you re-read the last couple of messages before replying, which I think encourages long thoughtful conversations that build on themselves.

I like the idea of treating **initial posts and comments as equally important**, to encourage these long conversations.  (As opposed to, say, Youtube comments, where each comment is a standalone response to the video without much interaction between comments).  So more of a chat-conversation-thread model instead of a post-and-short-comments model.

### Warrens and Plazas

Another dimension is **warren-vs-plaza**.  Warrens hide conversations in side rooms you have to click into, and plazas mix everything into one more public feeling space.

I'd predict that **warrens** feel more intimate and would encourage better conversations.  Examples: email threads, slack channels, chat groups in Signal.  This also helps confine annoying topics in their own warrens (like tech talk about SSB).  The downside is the more structured interface makes you click around to see into each warren, and posts might evolve clickbait subject lines to attract attention.

Patchwork is more of a **plaza** and I like it more than I thought I would.  It shows you the richness of conversations happening all around, and you can step in and join them when they become interesting.  It helps conversations gather more replies and last longer.  But it can become an overwhelming firehose when everyone's posting at once.

### Sun mode and Moon mode

In conclusion, imagine something like Github issues with dual ways of approaching the content: a structured list of threads, and a way to overhear the latest activity across all threads.

Mix and Zach have vocabulary for this:
%VzxV0Je3iaLmWEVCG2KqyvIZzFRwwfFTqXZLeD5at5g=.sha256

"☀️ Sun mode" and "🌘 Moon mode".  Sun mode is linear, purposeful, and directed.  Moon mode is coincidental, organic, wandering.  Maybe we could have both?

**Comment by ciex** [comment]
Thank you all so much for your thoughts <3 I am glad that some of you like the idea of branching out the timeline horizontally. I know this is quite experimental but I have been using it yesterday and today and somehow it just feels right to mee (except for some edge cases). I would post a video with some more scrolling here but I feel weird about putting people's scuttling on Github - so I encourage you to give it a try!

Regarding those edge cases: @christianbundy you asked about the layout if the comments end up requiring more height than the original post. My own intuition is that the "wrap-up" or aside should be limited in height to be able to fit the visible viewport height on a common 13 inch display if the original post is shorter than that. If the original post is longer than that the aside should not continue flowing past the original post.  We could either use Javascript to look at the rendered height of containers and react to that or do something by counting the length of the post's text content and ignore image height.

@cinnamon-bun I really like the ux perspective you are taking. I would add to that this critique of timelines for social media `%254UHRF%2FxbCaeyk%2FJgw92LBWXrfGm9SCo3YIfAVXwtNzw%3D.sha256`. (Side note: How can I make Oasis show me a post by giving it a message key? Putting it in search doesn't work for me so I just edited the URL, but that shouldn't be the way should it?)

Another experiment I would like to do is to create another kind of "popular" view that allows users to reflect on partciular time period (e.g. what was Scuttlebutt about in Winter 2019), I think I will write about that on ssb and link it here later.

I created a discussion issue about summaries (#221) as suggested by @christianbundy .

**Regarding this PR**: What is the process here? Should it wait for that discussion? Who decides whether this can go forward? I would love to polish it up if you can signal me that it would be merged eventually. I could also put in a config option to make it optional.

**Comment by ciex** [comment]
Oh by the way: Everything is way faster when Patchwork is running in the background.

**Comment by jedahan** [comment]
Generally our dev process has been: if it doesn't obviously *break* core features, we merge optimistically and then make improvements.

In that spirit the only thing I *might* consider broken is the 5x slowdown for popular. It would be a shame to block merging on just that, so either a global option, or turning it off just on the popular page would be enough for me to hit the big green merge button. Gonna try it out for the rest of the day.

**Comment by christianbundy** [comment]
> Generally our dev process has been: if it doesn't obviously *break* core features, we merge optimistically and then make improvements.

Right, specifically when the patch is "a minimal and accurate answer to exactly one identified and agreed problem". I'm going to optimistically merge this, let's see where it goes! Thanks for this patch @ciex.

**Comment by ciex** [comment]
Oh, now that was unexpected (ans also pretty cool). Actually, I would have liked to refactor some of things that I already commented on this PR. But I guess then I'll just make another PR for that.
